pkg/kubectl/cmd/delete.go

```diff
@@ -308,25 +323,91 @@ func DeleteResult(r *resource.Result, out io.Writer, ignoreNotFound bool, shortO
 	return nil
 }
 
-func cascadingDeleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper) error {
-	falseVar := false
-	deleteOptions := &metav1.DeleteOptions{OrphanDependents: &falseVar}
-	return deleteResource(info, out, shortOutput, mapper, deleteOptions)
-}
+// Deletes the given resource and then waits to confirm that the resource is deleted.
+func deleteResourceAndWait(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper, orphanDependents bool, waitForDeletion *bool, timeout time.Duration) error {
+	deleteOptions := &metav1.DeleteOptions{OrphanDependents: &orphanDependents}
+	obj, err := deleteResource(info, out, shortOutput, mapper, deleteOptions)
+	if err != nil {
+		return err
+	}
+	// WaitForDeletion is true by default except for namespace deletion.
+	if waitForDeletion == nil {
+		if info.Mapping.Resource == "namespaces" {
+			falseVar := false
+			waitForDeletion = &falseVar
+		} else {
+			trueVar := true
+			waitForDeletion = &trueVar
+		}
+	}
 
-func deleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper, deleteOptions *metav1.DeleteOptions) error {
-	if err := resource.NewHelper(info.Client, info.Mapping).DeleteWithOptions(info.Namespace, info.Name, deleteOptions); err != nil {
-		return cmdutil.AddSourceToErr("deleting", info.Source, err)
+	if !*waitForDeletion {
+		cmdutil.PrintSuccess(mapper, shortOutput, out, info.Mapping.Resource, info.Name, false, "deleted")
+		return nil
+	}
+	// Keep deleting the resource with uid precondition until we get a
+	// resource not found error.
+	// We use DELETE here instead of using GET to ensure that DELETE
+	// works without requiring GET permissions.
+	uid, err := getUID(obj)
+	if err != nil {
+		return fmt.Errorf("unexpected error in extracting uid: %s", err)
+	}
+	deleteOptions = &metav1.DeleteOptions{OrphanDependents: &orphanDependents}
+	if uid != "" {
```

Comment:
I thought we were going to simply return if we didn't have the uid info needed to do a conditional delete